File: Source/Extensions/Tessa.Extensions.Default.Client/Forums/HideForumTabUIExtension.cs
using System.Threading.Tasks;
using Tessa.Extensions.Default.Shared.Workflow.KrProcess;
using Tessa.Forums;
using Tessa.UI.Cards;
using Tessa.UI.Cards.Forms;

namespace Tessa.Extensions.Default.Client.Forums
{
    public class HideForumTabUIExtension : CardUIExtension
    {
        private readonly IKrTypesCache typesCache;
        public HideForumTabUIExtension(IKrTypesCache typesCache)
        {
            this.typesCache = typesCache;
        }

        public override async Task Initialized(ICardUIExtensionContext context)
        {
            ICardModel model = context.Model;
            if (model.MainFormWithTabs is not { } mainForm)
            {
                return;
            }

            KrComponents usedComponents = await KrComponentsHelper.GetKrComponentsAsync(model.Card, this.typesCache, context.CancellationToken);
            var krType = await KrProcessSharedHelper.TryGetKrTypeAsync(
                this.typesCache, model.Card, model.Card.TypeID, cancellationToken: context.CancellationToken);

            if (usedComponents.HasNot(KrComponents.UseForum) || 
                (krType != null && !krType.UseForum) ||
                (krType != null && krType.UseForum  && !krType.UseDefaultDiscussionTab))
            {
                if (model.Forms.TryGet(ForumHelper.ForumTabName, out var forumTab))
                {
                    // если вкладка была выбрана при том, что мы её удаляем, то мы, наверное, переходим из карточки-сателлита в основную карточку
                    // с восстановлением состояния основной карточки; это означает, что нам надо перевыбрать правильную вкладку по индексу
                    int selectedIndexToRestore = ReferenceEquals(mainForm.SelectedTab, forumTab)
                        ? model.Forms.IndexOf(forumTab)
                        : -1;
                    //forumTab.Blocks.RemoveAll(b => b.Name == ForumHelper.TopicsBlockName);

                    forumTab.IsCollapsed = true;
                    if (selectedIndexToRestore >= 0 && selectedIndexToRestore < model.Forms.Count)
                    {
                        mainForm.SelectedTab = model.Forms[selectedIndexToRestore];
                    }
                }
            }
        }
    }
}

File: WebClient SDK/src/default/forums/hideForumTabUIExtension.ts
import { CardUIExtension, ICardUIExtensionContext } from 'tessa/ui/cards';
import { DefaultFormMainViewModel } from 'tessa/ui/cards/forms';
import { getKrComponentsByCard, KrTypesCache, KrComponents } from 'tessa/workflow';
import { tryGetKrType } from '../workflow/krProcess/krUIHelper';
import { hasNotFlag } from 'tessa/platform';

export class HideForumTabUIExtension extends CardUIExtension {
  public initialized(context: ICardUIExtensionContext) {
    const model = context.model;
    const mainForm = model.mainForm;
    if (!(mainForm instanceof DefaultFormMainViewModel)) {
      return;
    }

    const usedComponents = getKrComponentsByCard(model.card, KrTypesCache.instance);
    const krType = tryGetKrType(KrTypesCache.instance, model.card, model.card.typeId);

    if (
      hasNotFlag(usedComponents, KrComponents.UseForum) ||
      (!!krType && !krType.useForum) ||
      (!!krType && krType.useForum && !krType.useDefaultDiscussionTab)
    ) {
      const forumTab = model.forms.find(x => x.name === 'Forum');
      if (forumTab) {
        forumTab.isCollapsed = true;
        mainForm.restoreSelectedTab();
      }
    }
  }
}
